scipy/interpolate/_polyint.py

```diff
@@ -541,33 +563,74 @@ class BarycentricInterpolator(_Interpolator1D):
+    >>> axs[0].legend(axs[0].get_lines(), labels, bbox_to_anchor=(0., 1.02, 1., .102),
+    ...               loc='lower left', ncols=3, mode="expand", borderaxespad=0., frameon=False)
+    >>> plt.tight_layout()
+    >>> plt.show()
     """
 
-    def __init__(self, xi, yi=None, axis=0):
-        _Interpolator1D.__init__(self, xi, yi, axis)
+    def __init__(self, xi, yi=None, axis=0, wi=None):
+        _Interpolator1DWithDerivatives.__init__(self, xi, yi, axis)
 
         self.xi = np.asfarray(xi)
         self.set_yi(yi)
         self.n = len(self.xi)
 
-        # See page 510 of Berrut and Trefethen 2004 for an explanation of the
-        # capacity scaling and the suggestion of using a random permutation of
-        # the input factors.
-        # At the moment, the permutation is not performed for xi that are
-        # appended later through the add_xi interface. It's not clear to me how
-        # to implement that and it seems that most situations that require
-        # these numerical stability improvements will be able to provide all
-        # the points to the constructor.
-        self._inv_capacity = 4.0 / (np.max(self.xi) - np.min(self.xi))
-        permute = np.random.permutation(self.n)
-        inv_permute = np.zeros(self.n, dtype=np.int32)
-        inv_permute[permute] = np.arange(self.n)
+        # cache derivative object to avoid re-computing the weights with every call.
+        self._diff_cij = None
 
-        self.wi = np.zeros(self.n)
-        for i in range(self.n):
-            dist = self._inv_capacity * (self.xi[i] - self.xi[permute])
-            dist[inv_permute[i]] = 1.0
-            self.wi[i] = 1.0 / np.prod(dist)
+        if wi is not None:
+            self.wi = wi
+        else:
+            # See page 510 of Berrut and Trefethen 2004 for an explanation of the
+            # capacity scaling and the suggestion of using a random permutation of
+            # the input factors.
+            # At the moment, the permutation is not performed for xi that are
+            # appended later through the add_xi interface. It's not clear to me how
+            # to implement that and it seems that most situations that require
+            # these numerical stability improvements will be able to provide all
+            # the points to the constructor.
+            self._inv_capacity = 4.0 / (np.max(self.xi) - np.min(self.xi))
+            permute = np.random.permutation(self.n)
+            inv_permute = np.zeros(self.n, dtype=np.int32)
+            inv_permute[permute] = np.arange(self.n)
+            self.wi = np.zeros(self.n)
+
+            for i in range(self.n):
+                dist = self._inv_capacity * (self.xi[i] - self.xi[permute])
+                dist[inv_permute[i]] = 1.0
+                prod = np.prod(dist)
+                if prod == 0.0:
+                    raise ValueError("Interpolation points xi must be distinct.")
```

Comment:
Can you add a test for this

Reply:
Just to check I understand - you mean a test which attempts to initialise a polynomial with some repeated interpolation points, and checks for the `ValueError`? If so, sure, I will do that.

Reply:
Exactly, just to make sure this error does get raised when you expect it to